scripts/trait-search-codegen.ts

```diff
@@ -0,0 +1,203 @@
+  } else if (type.getFlags() & ts.TypeFlags.Object) {
+    if (type.symbol?.getName() === 'Array') {
+      const elementType = checker.getTypeArguments(type as ts.TypeReference)[0];
+      return {
+        _tag: 'array',
+        elementType: traverseType(elementType)
+      } as ArrayNode;
+    } else {
+      const o = {} as any;
+      type.getProperties().forEach(prop => {
+        o[prop.name] = traverseType(checker.getTypeOfSymbolAtLocation(prop, prop.valueDeclaration!));
+      });
+      return o;
+    }
+  } else {
+    console.warn('Unknown type');
+  }
+  return { _tag: 'unknown' } as UnknownNode;
+};
+
+const accDatatableColumnsRecursive = (subtree: any): [string, LeafType][] => {
+  const out = [] as [string, LeafType][];
+  Object.entries(subtree).forEach(([k, v]) => {
+    if ('_tag' in (v as any)) {
+      out.push([k, (v as LeafType)]);
+    } else {
+      const leaves = accDatatableColumnsRecursive(v);
+      const mappedLeaves = leaves.map(([path, leaf]) => ([k + '.' + path, leaf] as [string, LeafType]));
+      out.push(...mappedLeaves);
+    }
+  });
+  return out;
+};
+
+// only supports string/null/array/obj values atm
+const valueToAST = (v: any, factory: ts.NodeFactory): ts.Expression => {
+  if (typeof v === 'string') {
+    return factory.createStringLiteral(v as string);
+  } else if (v === null) {
+    return factory.createNull();
+  } else if (Array.isArray(v)) {
+    return factory.createArrayLiteralExpression(
+      v.map(el => valueToAST(el, factory))
+    );
+  } else {
+    return factory.createObjectLiteralExpression(
+      Object.entries(v).map(([k, v2]) =>
+        factory.createPropertyAssignment(k, valueToAST(v2, factory)))
+    );
+  }
+};
+
+const generateExportObjectLiteral = (obj: any, name: string, path: string) => {
+  const factory = ts.factory;
+  const variableStatement = factory.createVariableStatement(
+    [factory.createModifier(ts.SyntaxKind.ExportKeyword)],
+    factory.createVariableDeclarationList([
+      factory.createVariableDeclaration(
+        name, undefined, undefined,
+        valueToAST(obj, factory)
+      ),
+    ], ts.NodeFlags.Const)
+  );
+  const sourceFile = ts.createSourceFile(path, '', ts.ScriptTarget.ESNext, false, ts.ScriptKind.TS);
+  const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });
+  const result = printer.printNode(ts.EmitHint.Unspecified, variableStatement, sourceFile);
+  fs.writeFileSync(path, `/* eslint-disable @typescript-eslint/quotes */\n/* eslint-disable max-len */\n/*\nGenerated file. Do not edit manually!\n*/\n\n${result}\n`);
+};
+
+const generateDatatableColumns = (tree: any) => {
+  const cols = accDatatableColumnsRecursive(tree);
+  console.log(cols);
+  generateExportObjectLiteral(cols, 'cols', './projects/laji/src/app/+trait-db/shared/trait-search/trait-search-table-columns.ts');
+};
+
+const generateFilters = (tree: any) => {
+  const cols = accDatatableColumnsRecursive(tree);
```

Comment:
Looks like `accDataTableColumnsRecursive` is being called first on `generateDatatableColumns` and then again on `generateFilters`. Would it be more effecient, if you would call it on main and give `cols` as param to the two functions?

Reply:
Performance doesn't matter here since it's not a script that is run regularly, but I fixed it anyway